contracts/staking/StakeHolder.sol

```diff
@@ -0,0 +1,266 @@
+        }
+        emit Distributed(msg.sender, msg.value, len);
+    }
+
+    /**
+     * @notice Get the balance of an account.
+     * @param _account The account to return the balance for.
+     * @return _balance The balance of the account.
+     */
+    function getBalance(address _account) external view returns (uint256 _balance) {
+        return balances[_account].stake;
+    }
+
+    /**
+     * @notice Determine if an account has ever staked.
+     * @param _account The account to determine if they have staked
+     * @return _everStaked True if the account has ever staked.
+     */
+    function hasStaked(address _account) external view returns (bool _everStaked) {
+        return balances[_account].hasStaked;
+    }
+
+    /**
+     * @notice Get the length of the stakers array.
+     * @dev This will be equal to the number of staker accounts that have ever staked.
+     *  Some of the accounts might have a zero balance, having staked and then
+     *  unstaked.
+     * @return _len The length of the stakers array.
+     */
+    function getNumStakers() external view returns (uint256 _len) {
+        return stakers.length;
+    }
+
+    /**
+     * @notice Get the staker accounts from the stakers array.
+     * @dev Given the stakers list could grow arbitrarily long. To prevent out of memory or out of
+     *  gas situations due to attempting to return a very large array, this function call specifies
+     *  the start offset and number of accounts to be return.
+     *  NOTE: This code will cause a panic if the start offset + number to return is greater than
+     *  the length of the array. Use getNumStakers before calling this function to determine the
+     *  length of the array.
+     * @param _startOffset First offset in the stakers array to return the account number for.
+     * @param _numberToReturn The number of accounts to return.
+     * @return _stakers A subset of the stakers array.
+     */
+    function getStakers(
+        uint256 _startOffset,
+        uint256 _numberToReturn
+    ) external view returns (address[] memory _stakers) {
+        address[] memory stakerPartialArray = new address[](_numberToReturn);
+        for (uint256 i = 0; i < _numberToReturn; i++) {
+            stakerPartialArray[i] = stakers[_startOffset + i];
+        }
+        return stakerPartialArray;
+    }
+
+    /**
+     * @notice Add more stake to an account.
+     * @dev If the account has a zero balance prior to this call, add the account to the stakers array.
+     * @param _account Account to add stake to.
+     * @param _amount The amount of stake to add.
+     * @param _existingAccountsOnly If true, revert if the account has never been used.
+     */
+    function _addStake(address _account, uint256 _amount, bool _existingAccountsOnly) private {
+        StakeInfo storage stakeInfo = balances[_account];
+        uint256 currentStake = stakeInfo.stake;
+        if (!stakeInfo.hasStaked) {
```

Comment:
in javascript/typescript world, if line 231 returned nothing (for a new account), like 233 would error. Interesting, it works fine in Solidity :)

Reply:
Yes, Solidity is different. Everything has a default, uninitialised value.